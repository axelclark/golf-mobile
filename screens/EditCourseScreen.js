import React from "react"
import { SafeAreaView, StyleSheet } from "react-native"
import { Mutation } from "react-apollo"
import gql from "graphql-tag"

import Header from "../components/Header"
import CourseForm from "../components/CourseForm"

const UPDATE_COURSE_MUTATION = gql`
  mutation($id: ID!, $course: CourseInput!) {
    course: updateCourse(id: $id, input: $course) {
      id
      numHoles
      name
      holes {
        id
        holeNumber
        par
      }
    }
  }
`

class EditCourseScreen extends React.Component {
  static navigationOptions = {
    title: "Edit Course",
  }

  constructor() {
    super()
    this.state = {
      data: {
        id: "",
        name: "",
        numHoles: "",
        holes: [],
      },
      errorMessage: "",
      nameError: "",
      numHolesError: "",
    }
  }

  componentWillMount() {
    const { navigation } = this.props
    let course = navigation.getParam("course", {})
    let holes = course.holes
    holes.forEach(hole => {
      delete hole.__typename
    })
    holes.sort((a, b) => a.holeNumber - b.holeNumber)
    course["holes"] = holes
    this.setState({ data: course })
  }

  handleFieldChange = field => value => {
    let data = { ...this.state.data }
    data[field] = value
    this.setState({ data })
  }

  onCompleted() {
    this.props.navigation.navigate("Courses")
  }

  handleError({ graphQLErrors, networkError }) {
    if (graphQLErrors) {
      graphQLErrors.map(({ message, details }) => {
        this.setState({ errorMessage: message })
        if (details) {
          if (details.name) {
            this.setState({ nameError: details.name.join() })
          }
          if (details.num_holes) {
            this.setState({ numHolesError: details.num_holes.join() })
          }
        }
      })
    }
    if (networkError) {
      this.setState({ errorMessage: networkError.message })
    }
  }

  render() {
    const { id, name, numHoles, holes } = this.state.data
    const updateParams = {
      id: id,
      course: {
        name: name,
        numHoles: Number(numHoles),
        holes: holes,
      },
    }
    return (
      <SafeAreaView styles={styles.container}>
        <Header title={"Edit Course"} />
        <Mutation
          mutation={UPDATE_COURSE_MUTATION}
          variables={updateParams}
          onCompleted={() => this.onCompleted()}
          onError={error => this.handleError(error)}
        >
          {updateCourseMutation => (
            <CourseForm
              {...this.state}
              handleSubmit={updateCourseMutation}
              handleFieldChange={this.handleFieldChange}
            />
          )}
        </Mutation>
      </SafeAreaView>
    )
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
})

export default EditCourseScreen
