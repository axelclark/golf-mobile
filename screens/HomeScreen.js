import React from 'react';
import {
  ScrollView,
  StyleSheet,
  Text,
  SafeAreaView,
  View,
} from 'react-native';

import { graphql } from 'react-apollo';
import gql from 'graphql-tag';

import Course from '../components/Course'

class HomeScreen extends React.Component {
  static navigationOptions = {
    header: null,
  };

  get courses() {
    const { data } = this.props;
    if (data && data.courses) {
      return data.courses;
    } else {
      return [];
    }
  }

  renderCourse(course) {
    return (
      <Course
        key={course.id}
        course={course}
      />
    )
  }

  render() {
    return (
      <SafeAreaView style={styles.container}>
        <ScrollView>
          <View style={styles.header}>
            <Text style={styles.headerText}>
              Courses
            </Text>
          </View>
          {this.courses.map(course => this.renderCourse(course))}
        </ScrollView>
      </SafeAreaView>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#f5f5f5',
  },
  header: {
    marginTop: 50,
    marginBottom: 50,
  },
  headerText: {
    textAlign: 'center',
    fontSize: 72,
    color: 'rgba(175, 47, 47, 0.25)',
    fontWeight: '100'
  },
});

const query = gql`
  { courses { id name numHoles} }
`;

export default graphql(query)(HomeScreen);
