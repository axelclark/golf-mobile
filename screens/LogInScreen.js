import React from "react"

import {
  AsyncStorage,
  SafeAreaView,
  ScrollView,
  StyleSheet,
  Switch,
  Text,
  View,
} from "react-native"

import {
  Button,
  FormLabel,
  FormInput,
  FormValidationMessage,
} from "react-native-elements"

import { Mutation } from "react-apollo"
import gql from "graphql-tag"

import Header from "../components/Header"
import Colors from "../constants/Colors"
import Sizes from "../constants/Sizes"
import Layout from "../constants/Layout"

const LOGIN_MUTATION = gql`
  mutation($email: String!, $password: String!) {
    session: login(email: $email, password: $password) {
      token
      user {
        email
      }
    }
  }
`

const CREATE_USER_MUTATION = gql`
  mutation($user: UserInput!) {
    session: createUser(input: $user) {
      token
      user {
        email
      }
    }
  }
`

const RESET_PASSWORD_MUTATION = gql`
  mutation($email: String!) {
    resetPassword(email: $email) {
      email
    }
  }
`

class LogInScreen extends React.Component {
  static navigationOptions = {
    header: null,
  }

  constructor() {
    super()
    this.state = {
      inputEmail: "",
      inputPassword: "",
      login: true,
      errorMessage: "",
      emailError: "",
      passwordError: "",
    }
  }

  updateEmail(input) {
    const inputEmail = input.toLowerCase()
    this.setState({ inputEmail })
  }

  updatePassword(inputPassword) {
    this.setState({ inputPassword })
  }

  resetErrors() {
    this.setState({
      errorMessage: "",
      emailError: "",
      passwordError: "",
    })
  }

  toggleLogIn(login) {
    this.resetErrors()
    this.setState({ login })
  }

  async saveToken({ session: { token } }) {
    try {
      await AsyncStorage.setItem("golf:userToken", JSON.stringify(token))
      this.props.navigation.navigate("AuthLoading")
    } catch (error) {
      // eslint-disable-next-line no-console
      console.log(`Error saving data ${error}`)
    }
  }

  handleError({ graphQLErrors, networkError }) {
    if (graphQLErrors) {
      graphQLErrors.map(({ message, details }) => {
        this.setState({ errorMessage: message })
        if (details) {
          if (details.email) {
            this.setState({ emailError: details.email.join() })
          }
          if (details.password) {
            this.setState({ passwordError: details.password.join() })
          }
        }
      })
    }
    if (networkError) {
      this.setState({ errorMessage: networkError.message })
    }
  }

  handleReset() {
    this.setState({ errorMessage: "Check email for link to reset password" })
  }

  render() {
    const {
      inputEmail,
      inputPassword,
      login,
      errorMessage,
      emailError,
      passwordError,
    } = this.state

    const logInParams = {
      email: inputEmail,
      password: inputPassword,
    }

    const resetParams = {
      email: inputEmail,
    }

    const registrationParams = {
      user: {
        email: inputEmail,
        password: inputPassword,
      },
    }

    return (
      <SafeAreaView style={styles.container}>
        <Header title={"ForeScore"} />
        <ScrollView>
          <View style={styles.formContainer}>
            <FormValidationMessage>{errorMessage}</FormValidationMessage>

            <FormLabel>Email</FormLabel>
            <FormInput
              onChangeText={text => this.updateEmail(text)}
              inputValue={inputEmail}
              textContentType={"emailAddress"}
              keyboardType={"email-address"}
              autoCapitalize={"none"}
            />
            <FormValidationMessage>{emailError}</FormValidationMessage>

            <FormLabel>Password</FormLabel>
            <FormInput
              onChangeText={text => this.updatePassword(text)}
              inputValue={inputPassword}
              textContentType={"password"}
              secureTextEntry={true}
            />
            <FormValidationMessage>{passwordError}</FormValidationMessage>

            <Mutation
              mutation={login ? LOGIN_MUTATION : CREATE_USER_MUTATION}
              variables={login ? logInParams : registrationParams}
              onCompleted={data => this.saveToken(data)}
              onError={error => this.handleError(error)}
            >
              {logInMutation => (
                <Button
                  title={login ? "Log In" : "Sign up"}
                  onPress={() => logInMutation()}
                  backgroundColor={
                    login ? Colors.darkPrimary : Colors.lightPrimary
                  }
                  style={styles.button}
                />
              )}
            </Mutation>

            {login && (
              <Mutation
                mutation={RESET_PASSWORD_MUTATION}
                variables={resetParams}
                onCompleted={() => this.handleReset()}
                onError={error => this.handleError(error)}
              >
                {resetPasswordMutation => (
                  <Button
                    title={"Reset Password"}
                    onPress={() => resetPasswordMutation()}
                    style={styles.button}
                    titleStyle={styles.resetButtonText}
                  />
                )}
              </Mutation>
            )}

            <View style={styles.switchContainer}>
              <View style={styles.textContainer}>
                <Text style={styles.text}>Sign Up</Text>
              </View>
              <Switch
                onValueChange={value => this.toggleLogIn(value)}
                value={login}
                ios_backgroundColor={Colors.lightPrimary}
                trackColor={{ true: Colors.darkPrimary }}
                style={styles.switch}
              />
              <View style={styles.textContainer}>
                <Text style={styles.text}>Log In</Text>
              </View>
            </View>
          </View>
        </ScrollView>
      </SafeAreaView>
    )
  }
}

const styles = StyleSheet.create({
  button: {
    marginTop: Sizes.large,
  },
  container: {
    flex: 1,
    backgroundColor: Colors.backgroundColor,
    justifyContent: "center",
    alignItems: "center",
  },
  formContainer: {
    flex: 1,
    maxWidth: Layout.window.width,
  },
  switch: {
    backgroundColor: Colors.darkPrimary,
  },
  switchContainer: {
    flexDirection: "row",
    marginTop: Sizes.large,
    justifyContent: "space-evenly",
  },
  text: {
    color: Colors.defaultLightText,
    fontSize: Sizes.medium,
  },
  textContainer: {
    justifyContent: "center",
  },
})

export default LogInScreen
