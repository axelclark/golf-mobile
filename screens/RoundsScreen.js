import React from 'react';
import {
  RefreshControl,
  SafeAreaView,
  ScrollView,
  StyleSheet,
  Text,
} from 'react-native';

import { Query } from 'react-apollo';
import gql from 'graphql-tag';

import RoundItem from '../components/RoundItem'
import Header from '../components/Header'
import Colors from '../constants/Colors'

export const ROUNDS_QUERY = gql`
  {
    rounds {
      id
      startedOn
      totalScore
      holesToPlay
      courseId
      course {
        id
        name
      }
    }
  }
`;

class RoundsScreen extends React.Component {
  static navigationOptions = {
    header: null,
  };

  constructor() {
    super()
    this.state = {
      refreshing: false,
    }
  }

  _onRefresh = (refetch) => {
    this.setState({refreshing: true});
    refetch().then(() => {
      this.setState({refreshing: false});
    });
  }

  renderRound(round) {
    return (
      <RoundItem
        key={round.id}
        round={round}
      />
    )
  }

  render() {
    return (
      <SafeAreaView style={styles.container}>
        <Header title={'Rounds'}/>
        <Query query={ROUNDS_QUERY}>
          {({ loading, error, data, refetch }) => {
            if (loading) return <Text>Fetching</Text>
              if (error) return <Text>Error</Text>
              const { rounds } = data
            return (
              <ScrollView
                refreshControl={
                  <RefreshControl
                    refreshing={this.state.refreshing}
                    onRefresh={() => this._onRefresh(refetch)}
                  />
                }
              >
                {rounds.map(round => this.renderRound(round))}
              </ScrollView>
            )
          }}
        </Query>
      </SafeAreaView>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: Colors.backgroundColor,
  },
});

export default RoundsScreen
