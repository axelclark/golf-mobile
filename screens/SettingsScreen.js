import React from "react"
import { Button } from "react-native-elements"
import { AsyncStorage, SafeAreaView, StyleSheet, Linking } from "react-native"

import { withApollo } from "react-apollo"

import Header from "../components/Header"
import Sizes from "../constants/Sizes"
import Colors from "../constants/Colors"

class SettingsScreen extends React.Component {
  static navigationOptions = {
    header: null,
  }

  async emailSupport() {
    const url = "mailto:forescoreapp@gmail.com"
    try {
      await Linking.openURL(url)
    } catch (error) {
      // eslint-disable-next-line no-console
      console.log(error)
    }
  }

  async logOutUser() {
    try {
      await AsyncStorage.removeItem("golf:userToken")
      await this.props.client.clearStore()
      this.props.navigation.navigate("AuthLoading")
    } catch (error) {
      // eslint-disable-next-line no-console
      console.log("error deleting token", error)
    }
  }

  render() {
    return (
      <SafeAreaView styles={styles.container}>
        <Header title={"Settings"} />
        <Button
          title="Log Out"
          buttonStyle={styles.button}
          onPress={() => this.logOutUser()}
        />
        <Button
          title="Email Support"
          buttonStyle={styles.button}
          onPress={() => this.emailSupport()}
        />
      </SafeAreaView>
    )
  }
}

const styles = StyleSheet.create({
  button: {
    marginTop: Sizes.large,
    backgroundColor: Colors.darkPrimary,
  },
  container: {
    flex: 1,
  },
})

export default withApollo(SettingsScreen)
