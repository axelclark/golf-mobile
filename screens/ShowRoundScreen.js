import React from 'react';
import {
  ScrollView,
  StyleSheet,
  Text,
  SafeAreaView,
  View,
} from 'react-native';

import { Query } from 'react-apollo';
import gql from 'graphql-tag';

import Score from '../components/Score'
import ScoresHeader from '../components/ScoresHeader'

class ShowRoundScreen extends React.Component {
  static navigationOptions = {
    title: 'Round',
  };

  sortByHole = scores => scores.sort((a, b) => {
    return parseInt(a.hole.holeNumber) - parseInt(b.hole.holeNumber)
  })

  render() {
    const id = this.props.navigation.getParam('id', 'Error')
    return (
      <SafeAreaView style={styles.container}>
        <Query query={ROUND_QUERY} variables={{ id }}>
          {({ loading, error, data }) => {
            if (loading) return <Text>Fetching</Text>
              if (error) return <Text>Error</Text>

              const scoresToRender = this.sortByHole(data.round.scores)

            return (
              <View>
                <ScoresHeader/>
                <ScrollView>
                  {scoresToRender.map(score => <Score key={score.id} score={score} />)}
                </ScrollView>
              </View>
            )
          }}
        </Query>
      </SafeAreaView>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#f5f5f5',
  },
});

export const ROUND_QUERY = gql`
  query ($id: ID!) {
    round(id: $id) {
      id
      startedOn
      course {
        id
        name
      },
      scores {
        id
        numStrokes
        hole {
          id
          holeNumber
          par
        }
      }
    }
  }
`
;

export default ShowRoundScreen
