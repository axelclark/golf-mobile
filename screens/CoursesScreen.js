import React from "react"
import {
  RefreshControl,
  ScrollView,
  StyleSheet,
  SafeAreaView,
  Text,
} from "react-native"

import { graphql, Query } from "react-apollo"
import gql from "graphql-tag"

import Course from "../components/Course"
import Header from "../components/Header"
import Colors from "../constants/Colors"

import { ROUNDS_QUERY } from "./RoundsScreen"

const COURSES_QUERY = gql`
  {
    courses {
      id
      name
      numHoles
    }
  }
`

class CoursesScreen extends React.Component {
  static navigationOptions = {
    header: null,
  }

  constructor() {
    super()
    this.state = {
      refreshing: false,
    }
  }

  _onRefresh = refetch => {
    this.setState({ refreshing: true })
    refetch().then(() => {
      this.setState({ refreshing: false })
    })
  }

  renderCourse(course) {
    return <Course key={course.id} course={course} />
  }

  render() {
    return (
      <SafeAreaView style={styles.container}>
        <Header title={"Courses"} />
        <Query query={COURSES_QUERY}>
          {({ loading, error, data, refetch }) => {
            if (loading) return <Text>Fetching</Text>
            if (error) return <Text>Error</Text>
            const { courses } = data
            return (
              <ScrollView
                refreshControl={
                  <RefreshControl
                    refreshing={this.state.refreshing}
                    onRefresh={() => this._onRefresh(refetch)}
                  />
                }
              >
                {courses.map(course => this.renderCourse(course))}
              </ScrollView>
            )
          }}
        </Query>
      </SafeAreaView>
    )
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: Colors.backgroundColor,
  },
})

export default graphql(ROUNDS_QUERY)(CoursesScreen)
