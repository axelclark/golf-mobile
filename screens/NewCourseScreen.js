import React from "react"
import { SafeAreaView, ScrollView, StyleSheet } from "react-native"
import { Mutation } from "react-apollo"
import gql from "graphql-tag"

import { COURSES_QUERY } from "../screens/CoursesScreen"
import Header from "../components/Header"
import CourseForm from "../components/CourseForm"

const CREATE_COURSE_MUTATION = gql`
  mutation($course: CourseInput!) {
    course: createCourse(input: $course) {
      id
      name
      numHoles
      holes {
        id
        par
        holeNumber
      }
    }
  }
`

class NewCourseScreen extends React.Component {
  static navigationOptions = {
    title: "New Course",
  }

  constructor() {
    super()
    this.state = {
      data: {
        name: "",
        numHoles: "",
        holes: [],
      },
      errorMessage: "",
      nameError: "",
      numHolesError: "",
    }
  }

  handleFieldChange = field => value => {
    let data = { ...this.state.data }
    data[field] = value
    this.setState({ data })
  }

  onCompleted({ course }) {
    this.props.navigation.navigate("EditCourse", { course })
  }

  handleError({ graphQLErrors, networkError }) {
    if (graphQLErrors) {
      graphQLErrors.map(({ message, details }) => {
        this.setState({ errorMessage: message })
        if (details) {
          if (details.name) {
            this.setState({ nameError: details.name.join() })
          }
          if (details.num_holes) {
            this.setState({ numHolesError: details.num_holes.join() })
          }
        }
      })
    }
    if (networkError) {
      this.setState({ errorMessage: networkError.message })
    }
  }

  render() {
    const { name, numHoles } = this.state.data
    const courseParams = {
      course: {
        name: name,
        numHoles: Number(numHoles),
      },
    }
    return (
      <SafeAreaView styles={styles.container}>
        <ScrollView scrollEnabled={false}>
          <Header title={"New Course"} />
          <Mutation
            mutation={CREATE_COURSE_MUTATION}
            variables={courseParams}
            onCompleted={data => this.onCompleted(data)}
            update={(store, { data: { course } }) => {
              const data = store.readQuery({ query: COURSES_QUERY })
              data.courses.push(course)
              data.courses.sort((a, b) => a.name.localeCompare(b.name))
              store.writeQuery({
                query: COURSES_QUERY,
                data,
              })
            }}
            onError={error => this.handleError(error)}
          >
            {createCourseMutation => (
              <CourseForm
                {...this.state}
                handleSubmit={createCourseMutation}
                handleFieldChange={this.handleFieldChange}
              />
            )}
          </Mutation>
        </ScrollView>
      </SafeAreaView>
    )
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
})

export default NewCourseScreen
