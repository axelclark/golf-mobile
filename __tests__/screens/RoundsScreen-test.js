import "react-native"
import React from "react"
import { MockedProvider } from "react-apollo/test-utils"
import TestRenderer from "react-test-renderer"
import NavigationTestUtils from "react-navigation/NavigationTestUtils"
import { ROUNDS_QUERY } from "../../screens/RoundsScreen"
import RoundsScreen from "../../screens/RoundsScreen"
import wait from "waait"

const mocks = [
  {
    request: {
      query: ROUNDS_QUERY,
    },
    result: {
      data: {
        rounds: [
          {
            id: "1",
            startedOn: "December",
            totalScore: "3",
            holesToPlay: "7",
            courseId: "2",
            course: {
              id: "2",
              name: "My Course",
            },
          },
        ],
      },
    },
  },
]

describe("Roundscreen snapshot", () => {
  it("renders RoundsScreen correctly", async () => {
    const component = TestRenderer.create(
      <MockedProvider mocks={mocks} addTypename={false}>
        <RoundsScreen />
      </MockedProvider>
    )
    await wait(0)
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it("renders RoundsScreen when loading", async () => {
    const component = TestRenderer.create(
      <MockedProvider mocks={mocks} addTypename={false}>
        <RoundsScreen />
      </MockedProvider>
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
