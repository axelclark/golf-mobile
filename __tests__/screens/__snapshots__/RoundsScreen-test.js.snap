// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Roundscreen snapshot renders RoundsScreen correctly 1`] = `
<RCTSafeAreaView
  style={
    Object {
      "backgroundColor": "hsl(0, 0%, 95%)",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "marginBottom": 48,
        "marginLeft": 8,
        "marginRight": 8,
        "marginTop": 48,
      }
    }
  >
    <Text
      adjustsFontSizeToFit={true}
      numberOfLines={1}
      style={
        Object {
          "color": "hsla(277, 100%, 33%, .5)",
          "fontSize": 96,
          "fontWeight": "100",
          "textAlign": "center",
        }
      }
    >
      Rounds
    </Text>
  </View>
</RCTSafeAreaView>
`;

exports[`Roundscreen snapshot renders RoundsScreen when loading 1`] = `
<RCTSafeAreaView
  style={
    Object {
      "backgroundColor": "hsl(0, 0%, 95%)",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "marginBottom": 48,
        "marginLeft": 8,
        "marginRight": 8,
        "marginTop": 48,
      }
    }
  >
    <Text
      adjustsFontSizeToFit={true}
      numberOfLines={1}
      style={
        Object {
          "color": "hsla(277, 100%, 33%, .5)",
          "fontSize": 96,
          "fontWeight": "100",
          "textAlign": "center",
        }
      }
    >
      Rounds
    </Text>
  </View>
  <Text>
    Fetching
  </Text>
</RCTSafeAreaView>
`;
